src/main/java/com/google/sps/database/RentPostDatabase.java

```diff
@@ -0,0 +1,28 @@
+package com.google.sps;
+
+import static com.google.sps.Constants.RENT_COLLECTION_NAME;
+
+import com.google.api.core.ApiFuture;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.FieldValue;
+import com.google.cloud.firestore.Firestore;
+import com.google.firebase.cloud.FirestoreClient;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.ServletException;
+
+/** Handles database interaction (add, fetch) for rent posts. */
+public class RentPostDatabase {
+  /**
+  * Add a rent post document to a collection using a map.
+  * @param post New rent post. 
+  */
+  public void addPost(RentPost post) {
+    Map<String, Object> postDocument = post.toMap();
+    FirestoreUtil.getDatabase()
+      .collection(RENT_COLLECTION_NAME)
+      .add(postDocument);
+  }
```

Comment:
What's the main intent of the separation between this class and the following? It potentially creates some odd deferred intialization costs. For instance, your servlets create a RentPostDatabase during init(), but the underlying database supplied by FirestoreUtil won't actually be initialized until the first call to addPost triggers creation, which is happening during a user-facing flow and may cause noticeable latency. 

A more flexible approach might be to have a static factory method for creating objects with a database interface, then have an implementation of that which uses Firestore. Then you could also separate your tests from firestore interface entirely.

Reply:
Changed database structure to a database interface, a firestore class that implements the database interface, and a database factory singleton class containing the sole reference to the database instance.